src/MergeContext.js

```diff
@@ -377,15 +381,131 @@ class MergeContext {
         return result;
     }
 
+    // Tries to load 'staging tag' for the PR.
+    async _loadTag() {
+       try {
+           if (this._tagSha)
+               return;
+           this._tagSha = await GH.getReference(this._stagingTag());
+       } catch (e) {
+           if (e.name === 'ErrorContext' && e.notFound())
+               Log.LogException(e, this._toString() + " " + this._stagingTag() + " not found");
+           else
+               throw e;
+       }
+    }
+
+    // Preconditions
+
+    // Check 'staging tag' state as merge condition.
+    // Returns true if there is a fresh merge commit with 'failure' status.
+    async _stagingFailed() {
+        await this._loadTag();
+        if (!this._tagSha)
+            return false;
+
+        const isFresh = await this._tagIsFresh();
+        this._log("staging tag is " + (isFresh ? "fresh" : "stale"));
+        if (isFresh) {
+            const commitStatus = await this._getStagingStatuses();
+            this._log("staging status details: " + commitStatus);
+            if (commitStatus.failed()) {
+                this._log("staging checks failed some time ago");
+                if (this._prMergeable() !== true)
+                    this._log("merge commit did not change due to conflicts with " + this._prBaseBranch());
+                return true;
+            }
+        }
+        if (!this._dryRun("deleting staging commit"))
+            await GH.deleteReference(this._stagingTag());
+        return false;
+    }
+
+    async _checkPreconditions() {
+        this._log("checking preconditions");
+
+        const pr = await GH.getPR(this._prNumber(), true);
+        // refresh PR data
+        assert(pr.number === this._prNumber());
+        this._rawPr = pr;
+
+        if (!this._prOpen()) {
+            this._logFailedCondition("open");
+            return StepResult.Fail();
+        }
+
+        if (await this._hasLabel(Config.mergedLabel(), this._prNumber())) {
+            this._logFailedCondition("already has merged status");
+            return StepResult.Fail();
+        }
+
+        if (this._prInProgress()) {
+            this._logFailedCondition("not in progress");
+            return StepResult.Fail();
+        }
+
+        if (!this._prMergeable()) {
+            this._logFailedCondition("mergeable");
+            return StepResult.Fail();
+        }
+
+        if (await this._stagingFailed()) {
```

Comment:
This is not a *pre*condition AFAICT -- staged PRs are handled _after_ they are staged (naturally).